go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -581,27 +586,29 @@ func (npw *nodePortWatcher) DeleteEndpoints(ep *kapi.Endpoints) {
 	}
 }
 
-func (npw *nodePortWatcher) UpdateEndpoints(old *kapi.Endpoints, new *kapi.Endpoints) {
-	name := ktypes.NamespacedName{Namespace: old.Namespace, Name: old.Name}
+func getEndpointAddresses(endpointSlice *discovery.EndpointSlice) []string {
+	endpointsAddress := make([]string, 0)
+	for _, endpoint := range endpointSlice.Endpoints {
+		endpointsAddress = append(endpointsAddress, endpoint.Addresses...)
+	}
+	return endpointsAddress
+}
 
-	if reflect.DeepEqual(new.Subsets, old.Subsets) {
+func (npw *nodePortWatcher) UpdateEndpointSlice(oldEpSlice, newEpSlice *discovery.EndpointSlice) {
+	oldEpAddr := getEndpointAddresses(oldEpSlice)
+	newEpAddr := getEndpointAddresses(newEpSlice)
+	if reflect.DeepEqual(oldEpSlice.Ports, newEpSlice.Ports) &&
+		reflect.DeepEqual(oldEpAddr, newEpAddr) {
 		return
 	}
 
-	klog.V(5).Infof("Updating endpoints %s in namespace %s", old.Name, old.Namespace)
-
-	// Delete old endpoint rules and add normal ones back
-	if len(new.Subsets) == 0 {
```

Comment:
I guess we can get away with removing this check because the [endpointSlice](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#endpoint-v1-discovery-k8s-io) API states 
that an internal Endpoint must contain at least one address 

```
addresses of this endpoint. The contents of this field are interpreted according to the corresponding EndpointSlice addressType field. Consumers must handle different types of addresses in the context of their own capabilities. This must contain at least one address but no more than 100.
```

Reply:
But we would still need to check if the  new endpointSlice has 0 endpoints right? In that scenario we should delete the rules on the node and add back the normal ones

Reply:
I can't remember the reason for this check, it was originally meant to be an "early exit" but I agree we should only care if the state of `hasHostNetworkEndpoints` changed here.